{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "767b6574",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import ast\n",
    "from pandas import json_normalize\n",
    "from Bio import SeqIO\n",
    "import csv\n",
    "import os\n",
    "from ete3 import NCBITaxa\n",
    "import numpy as np\n",
    "import sys\n",
    "import json\n",
    "import traceback\n",
    "import csv\n",
    "import glob\n",
    "import re\n",
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "920e14e9",
   "metadata": {},
   "source": [
    "## GC content table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ff995878",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gc = pd.read_csv('files/output/all__genomes_GC_content.tsv', sep='\\t')\n",
    "df_gc.reset_index(drop=True, inplace=True)\n",
    "\n",
    "df_gc['Genome_name'] = df_gc['Genome_name'].str.split('.').str[0]\n",
    "df_gc.rename(columns={'Genome_name': 'Genome_ID'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "669ef7c7",
   "metadata": {},
   "source": [
    "## Parsing of MGnify, JGI and HMP metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3cad4ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format GTDB taxonomy into the dictionary\n",
    "def format_taxonomy(value):\n",
    "    replacements = {'d': 'Domain', 'p': 'Phylum', 'c': 'Class', 'o': 'Order', 'f': 'Family', 'g': 'Genus', 's': 'Species'}\n",
    "    parts = value.split(';')\n",
    "\n",
    "    formatted_taxonomy = {}\n",
    "    for part in parts:\n",
    "        if '__' in part:\n",
    "            letter, taxonomy = part.split('__')\n",
    "            if letter in replacements:\n",
    "                formatted_part = {replacements[letter]: taxonomy}\n",
    "            else:\n",
    "                continue  # Skip unrecognized letters\n",
    "        else:\n",
    "            formatted_part = {'Unknown': part}\n",
    "        formatted_taxonomy.update(formatted_part)\n",
    "\n",
    "    return formatted_taxonomy\n",
    "\n",
    "# Extract taxonomy from dictionaries to columns\n",
    "def extract_taxon_info(row, key):\n",
    "    if row['Taxon'] is None:\n",
    "        return None\n",
    "    return row['Taxon'][key] if key in row['Taxon'] else None\n",
    "\n",
    "def taxid2names(taxid):\n",
    "    \"\"\"\n",
    "    Output example: 2: {'Bacteria', 1760: 'Actinomycetes'}\n",
    "    \"\"\"\n",
    "    taxid = int(taxid)\n",
    "    tax_name = ncbi.get_taxid_translator([taxid])[taxid]\n",
    "    lineage = ncbi.get_taxid_translator(ncbi.get_lineage(taxid))\n",
    "    x = list(lineage.values())\n",
    "    return lineage\n",
    "\n",
    "def rank2taxid(gene_taxid):\n",
    "    \"\"\"\n",
    "    Output example: {'superkingdom': 2, 'class': 1760}\n",
    "    \"\"\"\n",
    "    gene_lineage = ncbi.get_lineage(gene_taxid)\n",
    "    gene_lineage2ranks = ncbi.get_rank(gene_lineage)\n",
    "    gene_ranks2lineage = dict((rank, taxid) for (taxid, rank) in gene_lineage2ranks.items())\n",
    "    return gene_ranks2lineage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5b2d5bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_quality(row):\n",
    "    \"\"\"\n",
    "    Assign the quality level based on the N50 and the number of fasta records within 1 fna\n",
    "    \"\"\"\n",
    "    \n",
    "    if 'Isolate' in row['Genome_type']:\n",
    "        if row['Count'] > 5 or row['N50'] < 2500000:\n",
    "            return 'Isolate (fragmented)'\n",
    "        else:\n",
    "            return 'Isolate (complete)'\n",
    "    else:\n",
    "        return 'MAG'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5335e841",
   "metadata": {},
   "source": [
    "#### Sample classification table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f24eba1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Upload site classification table (old version)\n",
    "df_class = pd.read_csv('files/input/specimen_classifications/specimen_classification.tsv', delimiter='\\t')\n",
    "df_class['Original_HMP'] = df_class['Original_HMP'].str.lower()\n",
    "df_class['Original_JGI'] = df_class['Original_JGI'].str.lower()\n",
    "df_class['Original_HMP'] = df_class['Original_HMP'].str.split(',')\n",
    "df_class['Original_JGI'] = df_class['Original_JGI'].str.split(',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d7540e15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specimen</th>\n",
       "      <th>Category</th>\n",
       "      <th>Specimen_Type</th>\n",
       "      <th>Original_HMP</th>\n",
       "      <th>Original_JGI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gut_unclassified</td>\n",
       "      <td>Gut</td>\n",
       "      <td>Gut_unclassified</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[digestive system/host-associated, digestive s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gut_colon</td>\n",
       "      <td>Gut</td>\n",
       "      <td>Colon</td>\n",
       "      <td>[feces, stool]</td>\n",
       "      <td>[digestive system/huma fecal, digestive system...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gut_bile_duct</td>\n",
       "      <td>Gut</td>\n",
       "      <td>Bile_duct</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[digestive system/human bile duct]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Oral_unclassified</td>\n",
       "      <td>Oral</td>\n",
       "      <td>Oral_unclassified</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[digestive system/human oral]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Oral_plaque</td>\n",
       "      <td>Oral</td>\n",
       "      <td>Plaque</td>\n",
       "      <td>[attached keratinized gingiva, subgingival pla...</td>\n",
       "      <td>[digestive system/fossilized dental plaque]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Oral_mucosa</td>\n",
       "      <td>Oral</td>\n",
       "      <td>Mucosa</td>\n",
       "      <td>[buccal mucosa]</td>\n",
       "      <td>[none]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Oral_saliva</td>\n",
       "      <td>Oral</td>\n",
       "      <td>Saliva</td>\n",
       "      <td>[portion of saliva, saliva]</td>\n",
       "      <td>[none]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Oral_cavity</td>\n",
       "      <td>Oral</td>\n",
       "      <td>Oral_cavity</td>\n",
       "      <td>[hard palate, tongue dorsum]</td>\n",
       "      <td>[none]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Respiratory_unclassified</td>\n",
       "      <td>Respiratory</td>\n",
       "      <td>Respiratory_unclassified</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[respiratory system/human]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Respiratory_upper</td>\n",
       "      <td>Respiratory</td>\n",
       "      <td>Upper_respiratory</td>\n",
       "      <td>[anterior nares, palatine tonsils, throat]</td>\n",
       "      <td>[none]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Respiratory_lower</td>\n",
       "      <td>Respiratory</td>\n",
       "      <td>Lower_respiratory</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[respiratory system/human lung]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Urogenital_unclassified</td>\n",
       "      <td>Urogenital</td>\n",
       "      <td>Urogenital_unclassified</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[reproductive system/human]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Urogenital_vagina</td>\n",
       "      <td>Urogenital</td>\n",
       "      <td>Vagina</td>\n",
       "      <td>[mid vagina, vaginal introitus, posterior fornix]</td>\n",
       "      <td>[none]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Skin_unclassified</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Skin_unclassified</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[skin/human, skin/human skin]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Skin_head</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Head</td>\n",
       "      <td>[left retroauricular crease, right retroauricu...</td>\n",
       "      <td>[none]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Skin_arm</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Arm</td>\n",
       "      <td>[right antecubital fossa]</td>\n",
       "      <td>[none]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Specimen     Category             Specimen_Type                                       Original_HMP                                       Original_JGI\n",
       "0           Gut_unclassified          Gut          Gut_unclassified                                             [none]  [digestive system/host-associated, digestive s...\n",
       "1                  Gut_colon          Gut                     Colon                                     [feces, stool]  [digestive system/huma fecal, digestive system...\n",
       "2              Gut_bile_duct          Gut                 Bile_duct                                             [none]                 [digestive system/human bile duct]\n",
       "3          Oral_unclassified         Oral         Oral_unclassified                                             [none]                      [digestive system/human oral]\n",
       "4                Oral_plaque         Oral                    Plaque  [attached keratinized gingiva, subgingival pla...        [digestive system/fossilized dental plaque]\n",
       "5                Oral_mucosa         Oral                    Mucosa                                    [buccal mucosa]                                             [none]\n",
       "6                Oral_saliva         Oral                    Saliva                        [portion of saliva, saliva]                                             [none]\n",
       "7                Oral_cavity         Oral               Oral_cavity                       [hard palate, tongue dorsum]                                             [none]\n",
       "8   Respiratory_unclassified  Respiratory  Respiratory_unclassified                                             [none]                         [respiratory system/human]\n",
       "9          Respiratory_upper  Respiratory         Upper_respiratory         [anterior nares, palatine tonsils, throat]                                             [none]\n",
       "10         Respiratory_lower  Respiratory         Lower_respiratory                                             [none]                    [respiratory system/human lung]\n",
       "11   Urogenital_unclassified   Urogenital   Urogenital_unclassified                                             [none]                        [reproductive system/human]\n",
       "12         Urogenital_vagina   Urogenital                    Vagina  [mid vagina, vaginal introitus, posterior fornix]                                             [none]\n",
       "13         Skin_unclassified         Skin         Skin_unclassified                                             [none]                      [skin/human, skin/human skin]\n",
       "14                 Skin_head         Skin                      Head  [left retroauricular crease, right retroauricu...                                             [none]\n",
       "15                  Skin_arm         Skin                       Arm                          [right antecubital fossa]                                             [none]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_class.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9fdbc9e",
   "metadata": {},
   "source": [
    "### Mgnify"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95206d98",
   "metadata": {},
   "source": [
    "#### Parse MGnify gut"
   ]
  },
  {
   "cell_type": "raw",
   "id": "3a3bc3da",
   "metadata": {},
   "source": [
    "# Extract the number of records within each .fna file\n",
    "\n",
    "# !/bin/bash\n",
    "for i in ../raw_data/EBI/human-gut-v2-0-1/fna/*.fna; do \\\n",
    "    filename=\"${i%.fna}\"; \\\n",
    "    count=$(grep '>' $i | wc -l); \\\n",
    "    echo -e \"$filename\\t$count\"; done > files/output/mgnify_meta/mgnify_gut_fasta_counts.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "20ad78c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read MGnify gut metadata file\n",
    "df_mg_g = pd.read_csv('files/input/mgnify_meta/genomes-all_metadata_gut.tsv', delimiter='\\t')\n",
    "df_mg_g_counts = pd.read_csv('files/output/mgnify_meta/mgnify_gut_fasta_counts.tsv', delimiter = '\\t')\n",
    "\n",
    "# Select Genome/Lineage columns and rename them\n",
    "df_mg_g = df_mg_g[['Genome', 'Lineage', 'N50', 'Genome_type']]\n",
    "df_mg_g = pd.merge(df_mg_g, df_mg_g_counts, on=['Genome'], how='left')\n",
    "df_mg_g['Quality'] = df_mg_g.apply(compute_quality, axis=1)\n",
    "\n",
    "df_mg_g = df_mg_g.rename(columns={'Genome': 'Genome_ID', 'Lineage': 'Taxon'})\n",
    "df_mg_g['Taxon'] = df_mg_g['Taxon'].apply(format_taxonomy)\n",
    "\n",
    "# Create Body_site and Specimen_type columns\n",
    "df_mg_g['Body_site'] = 'Gut'\n",
    "df_mg_g['Specimen_type'] = 'Gut_unclassified'\n",
    "# Add Source column\n",
    "df_mg_g['Source'] = 'EBI'\n",
    "df_mg_g.drop(columns=['Count', 'Genome_type', 'N50'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "436c2263",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d600b0e0",
   "metadata": {},
   "source": [
    "#### Parse MGnify oral"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8817b9c9",
   "metadata": {},
   "source": [
    "# Extract the number of records within each .fna file\n",
    "\n",
    "for i in ../raw_data/EBI/human-oral-v1-0/fna/*.fna; do \\\n",
    "    filename=\"${i%.fna}\"; \\\n",
    "    count=$(grep '>' $i | wc -l); \\\n",
    "    echo -e \"$filename\\t$count\"; done > files/output/mgnify_meta/mgnify_oral_fasta_counts.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "93fb5485",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read MGnify oral metadata file\n",
    "df_mg_o = pd.read_csv('files/input/mgnify_meta/genomes-all_metadata_oral.tsv', delimiter='\\t')\n",
    "df_mg_o_counts = pd.read_csv('files/output/mgnify_meta/mgnify_oral_fasta_counts.tsv', delimiter = '\\t')\n",
    "\n",
    "df_mg_o = df_mg_o[['Genome', 'Lineage', 'N50', 'Genome_type']]\n",
    "df_mg_o = pd.merge(df_mg_o, df_mg_o_counts, on=['Genome'], how='left')\n",
    "\n",
    "df_mg_o['Quality'] = df_mg_o.apply(compute_quality, axis=1)\n",
    "\n",
    "df_mg_o = df_mg_o.rename(columns={'Genome': 'Genome_ID', 'Lineage': 'Taxon'})\n",
    "df_mg_o['Taxon'] = df_mg_o['Taxon'].apply(format_taxonomy)\n",
    "\n",
    "# Create Body_site and Specimen_type columns\n",
    "df_mg_o['Body_site'] = 'Oral'\n",
    "df_mg_o['Specimen_type'] = 'Oral_unclassified'\n",
    "# Add Source column\n",
    "df_mg_o['Source'] = 'EBI'\n",
    "df_mg_o.drop(columns=['Count', 'Genome_type', 'N50'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d163dca8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d9a4c37f",
   "metadata": {},
   "source": [
    "#### Add MGnify sample classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f1a3a38f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Modify sample classifications for MGnify data (final results taken from MGnify_meta_NEW.ipynb)\n",
    "sample_gut = pd.read_csv('files/output/mgnify_biosample_meta/mgnify_meta_sample_gut.tsv', sep='\\t')\n",
    "sample_gut.rename(columns={'Genome': 'Genome_ID'}, inplace=True)\n",
    "\n",
    "sample_oral = pd.read_csv('files/output/mgnify_biosample_meta/mgnify_meta_sample_oral.tsv', sep='\\t')\n",
    "sample_oral.rename(columns={'Genome': 'Genome_ID'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "97e5fb6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mgnify_gut = pd.merge(df_mg_g, sample_gut[['Genome_ID', 'Specimen']], on='Genome_ID', how='left')\n",
    "mgnify_gut['Specimen'].fillna(mgnify_gut['Specimen_type'], inplace=True)\n",
    "mgnify_gut.drop(columns = ['Body_site', 'Specimen_type'], axis=1, inplace=True)\n",
    "\n",
    "mgnify_oral = pd.merge(df_mg_o, sample_oral[['Genome_ID', 'Specimen']], on='Genome_ID', how='left')\n",
    "mgnify_oral['Specimen'].fillna(mgnify_oral['Specimen_type'], inplace=True)\n",
    "mgnify_oral.drop(columns = ['Body_site', 'Specimen_type'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c485ee32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cdf5bb9f",
   "metadata": {},
   "source": [
    "### HMP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56c87b71",
   "metadata": {},
   "source": [
    "#### Parse HMP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1704883d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read HMP metadata file\n",
    "df_hmp = pd.read_csv('files/input/hmp_meta/wgs_assembled_seq_set.tsv', delimiter='\\t')\n",
    "\n",
    "# Select id/body_site columns and rename them\n",
    "df_hmp = df_hmp[['id', 'body_site']]\n",
    "df_hmp = df_hmp.rename(columns={'id': 'Genome_ID', 'body_site': 'original_site'})\n",
    "df_hmp['original_site'] = df_hmp['original_site'].str.lower()\n",
    "\n",
    "df_hmp['Source'] = 'HMP'\n",
    "df_hmp['Quality'] = 'Metagenome'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02e48556",
   "metadata": {},
   "source": [
    "#### Add HMP sample classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e32793b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hmp['Specimen'] = pd.NA\n",
    "\n",
    "# Assign Specimen type and category (body site) from classification table to HMP data\n",
    "for index, row in df_hmp.iterrows():\n",
    "    original_site = row['original_site']\n",
    "    \n",
    "    for _, class_row in df_class.iterrows():\n",
    "        original_hmp = class_row['Original_HMP']\n",
    "        # Check if 'original_site' from HMP table is contained in the 'Original_HMP' list of classification table\n",
    "        if isinstance(original_hmp, list) and original_site in original_hmp:\n",
    "            df_hmp.at[index, 'Specimen'] = class_row['Specimen']\n",
    "\n",
    "# Remove original body site column\n",
    "df_hmp = df_hmp.drop('original_site', axis=1)\n",
    "df_hmp['Taxon'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8e92a257",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_ID</th>\n",
       "      <th>Source</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Specimen</th>\n",
       "      <th>Taxon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>54a24ca84a57a7d5b06687939f211398</td>\n",
       "      <td>HMP</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>Urogenital_vagina</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>54a24ca84a57a7d5b06687939f48f9f4</td>\n",
       "      <td>HMP</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>Oral_cavity</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>54a24ca84a57a7d5b06687939f386ad9</td>\n",
       "      <td>HMP</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>Skin_head</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>54a24ca84a57a7d5b06687939f2fa7f6</td>\n",
       "      <td>HMP</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>Oral_mucosa</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>54a24ca84a57a7d5b06687939f38fda1</td>\n",
       "      <td>HMP</td>\n",
       "      <td>Metagenome</td>\n",
       "      <td>Oral_mucosa</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Genome_ID Source     Quality           Specimen  \\\n",
       "0  54a24ca84a57a7d5b06687939f211398    HMP  Metagenome  Urogenital_vagina   \n",
       "1  54a24ca84a57a7d5b06687939f48f9f4    HMP  Metagenome        Oral_cavity   \n",
       "2  54a24ca84a57a7d5b06687939f386ad9    HMP  Metagenome          Skin_head   \n",
       "3  54a24ca84a57a7d5b06687939f2fa7f6    HMP  Metagenome        Oral_mucosa   \n",
       "4  54a24ca84a57a7d5b06687939f38fda1    HMP  Metagenome        Oral_mucosa   \n",
       "\n",
       "  Taxon  \n",
       "0  None  \n",
       "1  None  \n",
       "2  None  \n",
       "3  None  \n",
       "4  None  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hmp.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efc362df",
   "metadata": {},
   "source": [
    "### HMP reference genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cfbc6cb",
   "metadata": {},
   "source": [
    "#### Parse Bioproject PRJNA28331 metadata - Human Microbiome Project (HMP) Reference Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "374e4a42",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_39119/1479104618.py:129: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bioproject_filtered.drop(columns=['index'], inplace=True)\n",
      "/tmp/ipykernel_39119/1479104618.py:130: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bioproject_filtered['isolation_source'] = bioproject_filtered['isolation_source'].str.replace('\"', '')\n"
     ]
    }
   ],
   "source": [
    "# function for expanding rows with more than one GCF/GCA pair\n",
    "def expand_rows(row):\n",
    "    gcf_values = row['GCF_assembly_id']\n",
    "    gca_values = row['GCA_assembly_id']\n",
    "    num_values = max(len(gcf_values), len(gca_values))\n",
    "\n",
    "    rows = []\n",
    "    for i in range(num_values):\n",
    "        row_data = {\n",
    "            'biosample_id': row['biosample_id'],\n",
    "            'bioproject_id': row['bioproject_id'],\n",
    "            'isolation_source': row['isolation_source'],\n",
    "            'misc_param_hmp_supersite': row['misc_param_hmp_supersite'],\n",
    "            'misc_param_hmp_body_site': row['misc_param_hmp_body_site'],\n",
    "            'tissue': row['tissue']}\n",
    "\n",
    "        if i < len(gcf_values):\n",
    "            row_data['GCF_assembly_id'] = gcf_values[i]\n",
    "        else:\n",
    "            row_data['GCF_assembly_id'] = None\n",
    "\n",
    "        if i < len(gca_values):\n",
    "            row_data['GCA_assembly_id'] = gca_values[i]\n",
    "        else:\n",
    "            row_data['GCA_assembly_id'] = None\n",
    "\n",
    "        rows.append(row_data)\n",
    "\n",
    "    return pd.DataFrame(rows)\n",
    "\n",
    "with open('files/input/hmp_reference_meta/main_PRJNA28331.json') as file:\n",
    "    main_project = json.load(file)\n",
    "\n",
    "# Parse project metadata\n",
    "all = set()\n",
    "assembly_ids = []\n",
    "level = []\n",
    "biosample_ids = []\n",
    "bioproject_ids = []\n",
    "isolation_sources = []\n",
    "misc_param_hmp_supersites = []\n",
    "misc_param_hmp_body_sites = []\n",
    "tissue_values = []\n",
    "taxid = []\n",
    "for entry in main_project[\"reports\"]:   #fields of interest: 'assembly_level', 'bioproject_accession', 'biosample'\n",
    "        accession = entry['accession']\n",
    "        assembly_ids.append(accession)\n",
    "        assembly_level = entry['assembly_info']['assembly_level']\n",
    "        level.append(assembly_level)\n",
    "        biosample_ids.append(entry['assembly_info']['biosample']['accession'])\n",
    "        for projects in entry['assembly_info']['bioproject_lineage']:\n",
    "            for bioprojects in projects['bioprojects']:\n",
    "                if 'parent_accessions' in bioprojects and 'PRJNA28331' in bioprojects['parent_accessions']:\n",
    "                    bioproject_ids.append(bioprojects['accession'])\n",
    "\n",
    "        for attributes_key, attributes_value in entry['assembly_info']['biosample'].items():\n",
    "            if attributes_key == 'attributes':\n",
    "                transformed_dict = {}\n",
    "                for item in attributes_value:\n",
    "                    key = item['name']\n",
    "                    value = item['value']\n",
    "                    transformed_dict[key] = value\n",
    "\n",
    "                isolation_sources.append(transformed_dict.get('isolation_source'))\n",
    "                misc_param_hmp_supersites.append(transformed_dict.get('misc_param: HMP supersite'))\n",
    "                misc_param_hmp_body_sites.append(transformed_dict.get('misc_param: HMP body site'))\n",
    "                tissue_values.append(transformed_dict.get('tissue'))\n",
    "                taxid.append(transformed_dict.get('host_taxid'))\n",
    "\n",
    "data = {\n",
    "'assembly_id': assembly_ids,\n",
    "'level': level,\n",
    "'biosample_id': biosample_ids,\n",
    "'bioproject_id': bioproject_ids,\n",
    "'isolation_source': isolation_sources,\n",
    "'misc_param_hmp_supersite': misc_param_hmp_supersites,\n",
    "'misc_param_hmp_body_site': misc_param_hmp_body_sites,\n",
    "'tissue': tissue_values\n",
    "}\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "df.drop_duplicates(inplace=True)\n",
    "df['isolation_source'] = df['isolation_source'].replace('missing', None)\n",
    "df['misc_param_hmp_body_site'] = df['misc_param_hmp_body_site'].replace('not determined', None)\n",
    "df['misc_param_hmp_supersite'] = df['misc_param_hmp_supersite'].replace('not determined', None)\n",
    "df['misc_param_hmp_supersite'] = df['misc_param_hmp_supersite'].replace('other', None)\n",
    "pd.set_option('display.max_rows', 7000)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 5000)\n",
    "\n",
    "df_grouped = df.groupby(['biosample_id', 'bioproject_id']).agg({'assembly_id': list}).reset_index()\n",
    "df_grouped['GCF_assembly_id'] = df_grouped['assembly_id'].apply(lambda x: [a for a in x if a.startswith('GCF')]).apply(tuple)\n",
    "df_grouped['GCA_assembly_id'] = df_grouped['assembly_id'].apply(lambda x: [a for a in x if a.startswith('GCA')]).apply(tuple)\n",
    "df_grouped.drop(columns=['assembly_id'], inplace=True)\n",
    "\n",
    "bioproject_df = df_grouped.merge(df[['biosample_id', 'bioproject_id', 'isolation_source', 'misc_param_hmp_body_site', 'misc_param_hmp_supersite', 'tissue']],\n",
    "                                 on=['biosample_id', 'bioproject_id'], how='left')\n",
    "bioproject_df.drop_duplicates(inplace=True)\n",
    "bioproject_df.reset_index(inplace=True)\n",
    "bioproject_df.drop(columns=['index'], inplace=True)\n",
    "expanded_df = bioproject_df.apply(expand_rows, axis=1)\n",
    "bioproject_df_final = pd.concat(expanded_df.to_list(), ignore_index=True)\n",
    "\n",
    "# to remove the 'tissue' column, modify 1 row (since only 1 entry has values in this column), moving 'tissue'\n",
    "# value to the 'misc_param_hmp_supersite' column\n",
    "bioproject_df_final.loc[bioproject_df_final['GCF_assembly_id'] == 'GCF_000477535.1', 'misc_param_hmp_supersite'] = 'oral'\n",
    "bioproject_df_final = bioproject_df_final.merge(df[['assembly_id', 'level']], left_on='GCA_assembly_id',\n",
    "                                                right_on='assembly_id',\n",
    "                                                how='left')\n",
    "bioproject_df_final.drop(columns=['tissue', 'assembly_id'], inplace=True)\n",
    "bioproject_df_final = bioproject_df_final[['GCA_assembly_id', 'GCF_assembly_id', 'level',\n",
    "                                           'biosample_id', 'bioproject_id', 'isolation_source',\n",
    "                                           'misc_param_hmp_body_site', 'misc_param_hmp_supersite']]\n",
    "\n",
    "\n",
    "# Get a dataframe with rows containing no annotation at all\n",
    "unannotated_df = bioproject_df_final[pd.isnull(bioproject_df_final['isolation_source']) &\n",
    "                  pd.isnull(bioproject_df_final['misc_param_hmp_body_site']) &\n",
    "                  pd.isnull(bioproject_df_final['misc_param_hmp_supersite'])]\n",
    "\n",
    "\n",
    "# Filter the bioproject df to remove rows containing no annotation at all\n",
    "bioproject_filtered = bioproject_df_final[~(\n",
    "    pd.isnull(bioproject_df_final['isolation_source']) &\n",
    "    pd.isnull(bioproject_df_final['misc_param_hmp_body_site']) &\n",
    "    pd.isnull(bioproject_df_final['misc_param_hmp_supersite']))]\n",
    "\n",
    "bioproject_filtered.reset_index(inplace=True)\n",
    "bioproject_filtered.drop(columns=['index'], inplace=True)\n",
    "bioproject_filtered['isolation_source'] = bioproject_filtered['isolation_source'].str.replace('\"', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5413e297",
   "metadata": {},
   "source": [
    "#### Add HMP reference sample classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6c02ea1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "isolation_class = pd.read_csv('files/input/specimen_classifications/classifications_isolation_source.tsv', sep='\\t')\n",
    "hmp_body_class = pd.read_csv('files/input/specimen_classifications/classifications_hmp_bodysite.tsv', sep='\\t')\n",
    "hmp_super_class = pd.read_csv('files/input/specimen_classifications/classifications_hmp_supersite.tsv', sep='\\t')\n",
    "\n",
    "\n",
    "condition1 = (pd.notna(bioproject_filtered['isolation_source']))\n",
    "\n",
    "condition2 = (pd.isnull(bioproject_filtered['isolation_source']) &\n",
    "             pd.notna(bioproject_filtered['misc_param_hmp_body_site']))\n",
    "\n",
    "condition3 = (pd.isnull(bioproject_filtered['isolation_source']) &\n",
    "             pd.isnull(bioproject_filtered['misc_param_hmp_body_site']) &\n",
    "             pd.notna(bioproject_filtered['misc_param_hmp_supersite']))\n",
    "\n",
    "result_1 = bioproject_filtered[condition1].merge(isolation_class, on='isolation_source', how='left')\n",
    "result_2 = bioproject_filtered[condition2].merge(hmp_body_class, on='misc_param_hmp_body_site', how='left')\n",
    "result_3 = bioproject_filtered[condition3].merge(hmp_super_class, on='misc_param_hmp_supersite', how='left')\n",
    "\n",
    "final_result = pd.concat([result_1, result_2, result_3], ignore_index=True)\n",
    "\n",
    "# Manually replace annotations for PRJNA41955, PRJNA51495, PRJNA31017\n",
    "final_result.loc[final_result['bioproject_id'] == 'PRJNA41955', 'Specimen'] = 'Gut_colon'\n",
    "final_result.loc[final_result['bioproject_id'] == 'PRJNA51495', 'Specimen'] = 'Urogenital_vagina'\n",
    "final_result.loc[final_result['bioproject_id'] == 'PRJNA51495', 'Body_site'] = 'Urogenital'\n",
    "final_result.loc[final_result['bioproject_id'] == 'PRJNA31017', 'Specimen'] = 'Oral_plaque'\n",
    "\n",
    "# Filter based on Healthy/Disease sample status\n",
    "last_condition = final_result[final_result['Status'] != 'Disease']\n",
    "hmp_ref_samples = last_condition[(last_condition['Body_site'] != 'Unclassified')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f8556f",
   "metadata": {},
   "source": [
    "#### Add HMP reference taxonomy"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6c371e69",
   "metadata": {},
   "source": [
    "wget \"https://data.gtdb.ecogenomic.org/releases/latest/bac120_taxonomy.tsv\"\n",
    "wget \"https://data.gtdb.ecogenomic.org/releases/latest/ar53_taxonomy.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8c5b61f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Taxonomic lineage is taken from the latest version of GTDB (2023-06-09)\n",
    "gtdb_bac = pd.read_csv('files/input/hmp_reference_meta/bac120_taxonomy.tsv', sep='\\t', names=['assembly_ID', 'Taxonomy'])\n",
    "gtdb_ar = pd.read_csv('files/input/hmp_reference_meta/ar53_taxonomy.tsv', sep='\\t', names=['assembly_ID', 'Taxonomy'])\n",
    "gtdb_taxa = pd.concat([gtdb_bac, gtdb_ar])\n",
    "\n",
    "gtdb_taxa['assembly_ID'] = gtdb_taxa['assembly_ID'].str.replace(r'.*?(GCF|GCA)_', r'\\1_', regex=True)\n",
    "\n",
    "# Merge based on 'GCF_assembly_id', then on 'GCA_assembly_id' if 'Taxonomy' is NA\n",
    "merged_df1 = hmp_ref_samples.merge(gtdb_taxa, left_on='GCF_assembly_id', right_on='assembly_ID', how='left')\n",
    "missing_taxonomy = merged_df1[pd.isnull(merged_df1['Taxonomy'])]\n",
    "merged_df1 = merged_df1[pd.notna(merged_df1['Taxonomy'])]\n",
    "\n",
    "\n",
    "merged_df2 = missing_taxonomy.merge(gtdb_taxa, left_on='GCA_assembly_id', right_on='assembly_ID', how='left')\n",
    "merged_df2.drop(columns=['assembly_ID_x', 'Taxonomy_x'], inplace=True)\n",
    "\n",
    "\n",
    "bioproject_taxonomy = pd.concat([merged_df1, merged_df2], ignore_index=True)\n",
    "bioproject_taxonomy.drop(columns=['assembly_ID', 'assembly_ID_y'], inplace=True)\n",
    "\n",
    "bioproject_taxonomy['Taxonomy_z'] = bioproject_taxonomy['Taxonomy'].combine_first(bioproject_taxonomy['Taxonomy_y'])\n",
    "bioproject_taxonomy.drop(columns=['Taxonomy', 'Taxonomy_y'], inplace=True)\n",
    "bioproject_taxonomy = bioproject_taxonomy.rename(columns={'Taxonomy_z': 'Taxonomy'})\n",
    "\n",
    "\n",
    "# Manually add taxonomy annotations for 6 samples\n",
    "taxonomy_dict = {\n",
    "    'GCA_000401635.1': 'd__Eukaryota;p__Mucoromycota;c__Mucoromycetes;o__Mucorales;f__Mucoraceae;g__Mucor;s__Mucor circinelloides',\n",
    "    'GCA_000758865.1': 'd__Bacteria; p__Firmicutes; c__Tissierellia; o__; f__; g__; s__',\n",
    "    'GCA_000759145.1': 'd__Bacteria; p__Firmicutes; c__Tissierellia; o__Tissierellales; f__Peptoniphilaceae; g__Peptoniphilus; s__Peptoniphilus lacrimalis',\n",
    "    'GCA_001807715.1': 'd__Bacteria; p__Actinobacteria; c__Actinomycetia; o__Micrococcales; f__Micrococcaceae; g__Micrococcus; s__',\n",
    "    'GCA_001546155.1': 'd__Bacteria; p__Firmicutes; c__Tissierellia; o__Tissierellales; f__Peptoniphilaceae; g__Anaerococcus; s__Anaerococcus hydrogenalis',\n",
    "    'GCA_000220825.1': 'd__Bacteria; p__Fusobacteria; c__Fusobacteriia; o__Fusobacteriales; f__Fusobacteriaceae; g__Fusobacterium; s__Fusobacterium nucleatum'\n",
    "}\n",
    "\n",
    "for gca_id, taxonomy in taxonomy_dict.items():\n",
    "    bioproject_taxonomy.loc[bioproject_taxonomy['GCA_assembly_id'] == gca_id, 'Taxonomy'] = taxonomy\n",
    "\n",
    "    \n",
    "bioproject_taxonomy['Taxonomy'] = bioproject_taxonomy['Taxonomy'].apply(format_taxonomy)\n",
    "\n",
    "# keys_to_extract = ['Domain', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species']\n",
    "\n",
    "# # Parse the lineage into separate columns\n",
    "# for key in keys_to_extract:\n",
    "#     bioproject_taxonomy[key] = bioproject_taxonomy['Taxonomy'].apply(lambda x: x.get(key, None))\n",
    "\n",
    "# bioproject_taxonomy.drop(columns=['Taxonomy'], inplace=True)\n",
    "\n",
    "bioproject_taxonomy = bioproject_taxonomy.rename(columns={'level': 'Quality', 'Taxonomy': 'Taxon'})\n",
    "\n",
    "\n",
    "# Add Quality column, add Source column, perform final filtering\n",
    "bioproject_taxonomy['Source'] = 'NCBI_PRJNA28331'\n",
    "bioproject_taxonomy['Quality'] = bioproject_taxonomy['Quality'].replace({'Chromosome': 'Isolate (complete)', 'Complete Genome': 'Isolate (complete)',\n",
    "                                              'Contig': 'Isolate (fragmented)', 'Scaffold': 'Isolate (fragmented)'})\n",
    "\n",
    "\n",
    "bioproject_taxa_filtered = bioproject_taxonomy[~bioproject_taxonomy['Body_site'].str.contains('Joint|Eye', case=False, na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8bfcbda1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_39119/91954205.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  hmp_reference['Genome_ID'] = hmp_reference['GCA_assembly_id'].str.split('.').str[0]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_ID</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Source</th>\n",
       "      <th>Specimen</th>\n",
       "      <th>GC%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000174135.1</td>\n",
       "      <td>Isolate (fragmented)</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Bacillota', ...</td>\n",
       "      <td>NCBI_PRJNA28331</td>\n",
       "      <td>Skin_arm</td>\n",
       "      <td>32.769365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000174175.1</td>\n",
       "      <td>Isolate (fragmented)</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Campylobacte...</td>\n",
       "      <td>NCBI_PRJNA28331</td>\n",
       "      <td>Oral_plaque</td>\n",
       "      <td>44.849322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000159235.2</td>\n",
       "      <td>Isolate (fragmented)</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Actinomyceto...</td>\n",
       "      <td>NCBI_PRJNA28331</td>\n",
       "      <td>Urogenital_vagina</td>\n",
       "      <td>42.656827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000159375.2</td>\n",
       "      <td>Isolate (fragmented)</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Bacillota', ...</td>\n",
       "      <td>NCBI_PRJNA28331</td>\n",
       "      <td>Gut_unclassified</td>\n",
       "      <td>43.736761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000164155.2</td>\n",
       "      <td>Isolate (complete)</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Actinomyceto...</td>\n",
       "      <td>NCBI_PRJNA28331</td>\n",
       "      <td>Skin_unclassified</td>\n",
       "      <td>68.765442</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Genome_ID               Quality                                              Taxon           Source           Specimen        GC%\n",
       "0  GCA_000174135.1  Isolate (fragmented)  {'Domain': 'Bacteria', 'Phylum': 'Bacillota', ...  NCBI_PRJNA28331           Skin_arm  32.769365\n",
       "1  GCA_000174175.1  Isolate (fragmented)  {'Domain': 'Bacteria', 'Phylum': 'Campylobacte...  NCBI_PRJNA28331        Oral_plaque  44.849322\n",
       "2  GCA_000159235.2  Isolate (fragmented)  {'Domain': 'Bacteria', 'Phylum': 'Actinomyceto...  NCBI_PRJNA28331  Urogenital_vagina  42.656827\n",
       "3  GCA_000159375.2  Isolate (fragmented)  {'Domain': 'Bacteria', 'Phylum': 'Bacillota', ...  NCBI_PRJNA28331   Gut_unclassified  43.736761\n",
       "4  GCA_000164155.2    Isolate (complete)  {'Domain': 'Bacteria', 'Phylum': 'Actinomyceto...  NCBI_PRJNA28331  Skin_unclassified  68.765442"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmp_reference = bioproject_taxa_filtered[['GCA_assembly_id', 'Quality', 'Taxon', 'Source', 'Specimen']]\n",
    "hmp_reference['Genome_ID'] = hmp_reference['GCA_assembly_id'].str.split('.').str[0]\n",
    "hmp_ref_gc = pd.merge(hmp_reference, df_gc, on='Genome_ID', how='left')\n",
    "hmp_ref_gc.drop(columns=['Genome_ID'], inplace=True)\n",
    "hmp_ref_gc = hmp_ref_gc.rename(columns={'GCA_assembly_id': 'Genome_ID'})\n",
    "hmp_ref_gc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d89d153",
   "metadata": {},
   "source": [
    "### JGI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df8f12b7",
   "metadata": {},
   "source": [
    "#### Parse JGI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f981c3b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read JGI metadata file\n",
    "df_jgi = pd.read_csv('files/input/jgi_meta/genome_metadata_human.tsv', delimiter='\\t')\n",
    "\n",
    "# Select genome_id, ecosystem, ecosystem_type, habitat columns and rename them\n",
    "df_jgi = df_jgi[['genome_id', 'ecosystem', 'ecosystem_type', 'habitat']]\n",
    "df_jgi['original_site'] = df_jgi.apply(lambda row: f\"{row['ecosystem_type']}/{row['habitat']}\", axis=1)\n",
    "df_jgi = df_jgi.drop('ecosystem_type', axis=1)\n",
    "df_jgi = df_jgi.drop('habitat', axis=1)\n",
    "\n",
    "df_jgi = df_jgi.rename(columns={'genome_id': 'Genome_ID', 'ecosystem': 'Taxon'})\n",
    "df_jgi['Taxon'] = df_jgi['Taxon'].apply(format_taxonomy)\n",
    "df_jgi['original_site'] = df_jgi['original_site'].str.lower()\n",
    "\n",
    "df_jgi['Source'] = 'JGI'\n",
    "df_jgi['Quality'] = 'MAG'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f895f278",
   "metadata": {},
   "source": [
    "#### Add JGI sample classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8ff411a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_jgi['Specimen'] = pd.NA\n",
    "\n",
    "# Assign Specimen type and category (body site) from classification table to JGI data\n",
    "for index, row in df_jgi.iterrows():\n",
    "    original_site = row['original_site']\n",
    "    \n",
    "\n",
    "    for _, class_row in df_class.iterrows():\n",
    "        original_jgi = class_row['Original_JGI']\n",
    "         # Check if 'original_site' from JGI table is contained in the 'Original_JGI' list of classification table\n",
    "        if isinstance(original_jgi, list) and original_site in original_jgi:\n",
    "            df_jgi.at[index, 'Specimen'] = class_row['Specimen']\n",
    "# Remove original body site column\n",
    "df_jgi = df_jgi.drop('original_site', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5098153f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_ID</th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Source</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Specimen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300007803_15</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Firmicutes_C...</td>\n",
       "      <td>JGI</td>\n",
       "      <td>MAG</td>\n",
       "      <td>Gut_unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300007803_19</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Campylobacte...</td>\n",
       "      <td>JGI</td>\n",
       "      <td>MAG</td>\n",
       "      <td>Gut_unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300007803_10</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Bacteroidota...</td>\n",
       "      <td>JGI</td>\n",
       "      <td>MAG</td>\n",
       "      <td>Gut_unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300007803_26</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Patescibacte...</td>\n",
       "      <td>JGI</td>\n",
       "      <td>MAG</td>\n",
       "      <td>Gut_unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300007803_9</td>\n",
       "      <td>{'Domain': 'Bacteria', 'Phylum': 'Firmicutes',...</td>\n",
       "      <td>JGI</td>\n",
       "      <td>MAG</td>\n",
       "      <td>Gut_unclassified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Genome_ID                                              Taxon Source Quality          Specimen\n",
       "0  3300007803_15  {'Domain': 'Bacteria', 'Phylum': 'Firmicutes_C...    JGI     MAG  Gut_unclassified\n",
       "1  3300007803_19  {'Domain': 'Bacteria', 'Phylum': 'Campylobacte...    JGI     MAG  Gut_unclassified\n",
       "2  3300007803_10  {'Domain': 'Bacteria', 'Phylum': 'Bacteroidota...    JGI     MAG  Gut_unclassified\n",
       "3  3300007803_26  {'Domain': 'Bacteria', 'Phylum': 'Patescibacte...    JGI     MAG  Gut_unclassified\n",
       "4   3300007803_9  {'Domain': 'Bacteria', 'Phylum': 'Firmicutes',...    JGI     MAG  Gut_unclassified"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_jgi.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f0b3b08",
   "metadata": {},
   "source": [
    "## Merge all dataframes to obtain a full metadata table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "cebf5e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merged_hmp_jgi = pd.concat([df_hmp, df_jgi], axis=0)\n",
    "merged_ebi_hmp_jgi = pd.concat([mgnify_gut, mgnify_oral, df_hmp, df_jgi], axis=0)\n",
    "\n",
    "# Add GC content\n",
    "metadata_table = pd.merge(merged_ebi_hmp_jgi, df_gc, on='Genome_ID', how='left')\n",
    "\n",
    "final_df = pd.concat([metadata_table, hmp_ref_gc], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f4bc43b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "keys_to_extract = ['Domain', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species']\n",
    "for key in keys_to_extract:\n",
    "    final_df[key] = final_df['Taxon'].apply(lambda x: x.get(key) if pd.notna(x) else None)\n",
    "\n",
    "final_df.drop(columns=['Taxon'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7dda24ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = final_df.fillna('NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5530df86",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2ab8b1f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv('files/output/ABC_HuMi_all_metadata_latest.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcf1313f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
